### ISSUE TITLE
gaugan ecnoder not training

### ISSUE THREAD

**Issue by yonigottesman** [soumik12345/tf2_gans#14]
Hi in the keras tutorial and in this code:
https://github.com/soumik12345/tf2_gans/blob/master/gaugan/models/gaugan.py
The encoder trainable_variables never get updated

**Comment by soumik12345** [comment]
HI @yonigottesman, the encoder trainable variables are updated when we train the `self.combined_model` as you can check here: https://github.com/soumik12345/tf2_gans/blob/master/gaugan/models/gaugan.py#L149

**Comment by yonigottesman** [comment]
the encoder is not part of the combined_model

**Comment by sayakpaul** [comment]
Thanks for reporting it. We will look into it and update th thread.

**Comment by yonigottesman** [comment]
Also this implementation looks 99% like the one in 
https://github.com/PacktPublishing/Hands-On-Image-Generation-with-TensorFlow-2.0/blob/master/Chapter06/ch6_gaugan.ipynb
maybe you should add some note that its based on it

**Comment by soumik12345** [comment]
**TODOs:**
- [ ] Add necessary acknowledgements
- [ ] Updated the encoder trainable variables

**Comment by sayakpaul** [comment]
I am still skeptical about how the encoder variables should be updated. In this [example](https://github.com/PacktPublishing/Hands-On-Image-Generation-with-TensorFlow-2.0/blob/master/Chapter06/ch6_gaugan.ipynb) there are also no direct updates there. The chain is -> encoder(images) -> sample_from_mean_variance -> pass to the generator. It's also not under `tf.GradientTape()`, so its variables are not being tracked anyway. This part is still confusing and definitely does not get clarified from the paper figures.

**Comment by yonigottesman** [comment]
@sayakpaul you cannot do `apply_gradients` on `self.encoder.trainable_variables` if you didn't first compute the gradients using `gradients = tape.gradient`
when you did `zip(gradients, self.combined_model.trainable_variables + self.encoder.trainable_variables)` the length of the lists are not the same and zip takes only the `self.combined_model.trainable_variables` part

**Comment by sayakpaul** [comment]
I understand that but I believe my comment was more about updating the encoder params in the first place.

**Comment by yonigottesman** [comment]
the example from the book is also wrong

**Comment by yonigottesman** [comment]
@sayakpaul why do you have a problem with:
`gradients = tape.gradient(total_loss, self.combined_model.trainable_variables + self.encoder.trainable_variables)`
and then 
`self.generator_optimizer.apply_gradients(zip(gradients, self.combined_model.trainable_variables + self.encoder.trainable_variables)`

**Comment by sayakpaul** [comment]
**_No problems_** with that. I just have confusion on whether to update the encoder params (which ideally we should). This became more apparent after looking at the Packt book example.
